ext/storage_sources/s3_store/s3_storage_source.cpp

```diff
@@ -845,51 +783,53 @@ S3Flush(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session, WT_FILE_SYSTEM *f
         return (ENOENT);
     }
 
-    /* Upload the object into the bucket. */
+    // Upload the object into the bucket.
     if (ret = (fs->connection->PutObject(object, source)) != 0)
         s3->log->LogErrorMessage("S3Flush: PutObject request to S3 failed.");
+    else
+        s3->log->LogDebugMessage("S3Flush: Uploaded object to S3.");
 
     return (ret);
 }
 
-/*
- * S3FlushFinish --
- *     Flush local file to cache.
- */
+// Flush local file to cache.
 static int
 S3FlushFinish(WT_STORAGE_SOURCE *storage, WT_SESSION *session, WT_FILE_SYSTEM *fileSystem,
   const char *source, const char *object, const char *config)
 {
-    S3_STORAGE *s3 = (S3_STORAGE *)storage;
-    S3_FILE_SYSTEM *fs = (S3_FILE_SYSTEM *)fileSystem;
-    /* Constructing the pathname for source and cache from file system and local.  */
+    S3Storage *s3 = (S3Storage *)storage;
+    S3FileSystem *fs = (S3FileSystem *)fileSystem;
+    // Constructing the pathname for source and cache from file system and local.
     std::string srcPath = S3Path(fs->homeDir, source);
     std::string destPath = S3Path(fs->cacheDir, object);
 
-    /* Linking file with the local file. */
-    int ret = link(srcPath.c_str(), destPath.c_str());
+    // Linking file with the local file.
+    std::error_code ec;
+    std::filesystem::create_hard_link(srcPath.c_str(), destPath.c_str(), ec);
+    int ret = ec.value();
     if (ret != 0) {
         ret = errno;
         s3->log->LogErrorMessage(
           "S3FlushFinish: link " + std::string(source) + " to " + destPath + " failed");
         return (ret);
     }
 
-    /* The file should be read-only. */
-    ret = chmod(destPath.c_str(), 0444);
+    // The file should be read-only.
+    std::filesystem::permissions(destPath.c_str(),
+      std::filesystem::perms::owner_read | std::filesystem::perms::group_read |
+        std::filesystem::perms::others_read,
+      std::filesystem::perm_options::add, ec);
+    ret = ec.value();
     if (ret != 0) {
         ret = errno;
```

Comment:
Are we sure that std::filesystem::create_hard_link sets an errno?

Reply:
From the documentation:
The overload taking a [std::error_code](http://en.cppreference.com/w/cpp/error/error_code)& parameter sets it to the OS API error code if an OS API call fails, and executes ec.clear() if no errors occur.

Where ec.clear() sets the value to 0.
I used error_code here to keep the ret logic. We could use a try-catch statement here if that would be more informative?